crates/ark/src/interface.rs

```diff
@@ -890,18 +894,25 @@ impl RMain {
         if let Some(req) = std::mem::take(&mut self.active_request) {
             // FIXME: Race condition between the comm and shell socket threads.
             //
-            // Send info for the next prompt to frontend. This handles
-            // custom prompts set by users, e.g. `options(prompt = ,
-            // continue = )`, as well as debugging prompts, e.g. after a
-            // call to `browser()`.
-            let event = UiFrontendEvent::PromptState(PromptStateParams {
-                input_prompt: info.input_prompt.clone(),
-                continuation_prompt: info.continuation_prompt.clone(),
+            // Perform a refresh of the frontend state
+            // (Prompts, working directory, etc)
+            self.with_mut_ui_comm_tx(|ui_comm_tx| {
+                let input_prompt = info.input_prompt.clone();
+                let continuation_prompt = info.continuation_prompt.clone();
+
+                ui_comm_tx.send_refresh(input_prompt, continuation_prompt);
             });
-            {
-                let kernel = self.kernel.lock().unwrap();
-                kernel.send_ui_event(event);
-            }
+
+            // Check for pending graphics updates
+            // (Important that this occurs while in the "busy" state of this ExecuteRequest
+            // so that the `parent` message is set correctly in any Jupyter messages)
+            unsafe {
+                graphics_device::on_did_execute_request(
+                    self.comm_manager_tx.clone(),
+                    self.iopub_tx.clone(),
+                    self.is_ui_comm_connected() && self.session_mode == SessionMode::Console,
+                )
+            };
```

Comment:
Previously we did a "prompt" refresh here right before unblocking shell, but over on shell when we got unblocked we then did a working directory refresh and a graphics device refresh.

We now do _all_ 3 of those here, which seems way more appropriate